### ISSUE TITLE
Best Approach Given Data Set Size - Pytorch Explainer

### ISSUE THREAD

**Issue by jlevy44** [shap/shap#374]
Hey there!

I'm working on a project with 9000 samples and 300k features. 

These features were used in multi-class classification and regression (multi-output) problems, using a Pytorch neural network model.

I've been trying to find the shapley values for the test set (~20% of the data), but this seems to be a computationally intractable problem, as I'm either getting huge CUDA or CPU memory errors or too lengthy compute times when I try to get the shaply_values after using Kernel, Gradient or Deep Explainers.

So far I've tried to split the shap samples up (different from data set samples) for gradient/kernel explainers and take the mean shapley values to try to handle large compute memory, and am considering running shap.kmeans beforehand to reduce the number of background samples. However, with 300k features, wouldn't kmeans center close to zero and the new background samples be meaningless? Perhaps I am mistaken about how it is used...

What would you recommend to handle this computationally intense problem in any reasonable amount of time? 

Are the deep or gradient explainers worth using in this case?

Looking to get array of ~3000 samples x 300k features of shapley values as end product (test set).

Thanks! This is an awesome library and looking forward to including it in my research in the years ahead if it works.

**Comment by slundberg** [comment]
Hey! I think the most scalable one right for PyTorch is the GradientExplainer. What happens if you try and explain just 10 samples using GradientExplainer? If you get memory errors you can use the `batch_size` argument to control how many samples get batched.

**Comment by jlevy44** [comment]
I believe the memory is fine. But seems like it’ll take too long to compute. What can I do to speed this up?

Thanks!

**Comment by jlevy44** [comment]
I have already tried what you mentioned, which I believe addressed the memory issues. I ran the jobs on a batch system, and they were taking too long to complete so I ended up terminating them. Just need to optimize for speed.

**Comment by jlevy44** [comment]
Oh I see what you mean. Yeah I’ll try just 10 samples and see what happens. In any case, I could probably break up my test array and concatenation the Shapley values at the end.

**Comment by jlevy44** [comment]
If you have a list of recommendations to optimize the algorithm for speed and memory, though I’m now aware of a few ways, that would be great to know everything I can try!

**Comment by slundberg** [comment]
One other thing is how many outputs do you have? If you have more than a couple you could use the `ranked_ouputs` option to only explain the top output.

Other than that I am not sure without playing with it myself. The `_PyTorchGradientExplainer` class is only 100 lines of python/pytorch code, so you could walk through it to check the performance and see where it is taking its time. By default it should just be running 200 back-propagation passes per sample and output you are explaining.

**Comment by jlevy44** [comment]
The number of outputs is variable, but let's say about 50. It is important to be able to explain all outputs though for my use case.

Thanks for the help! Hope this helps out others looking to make explanations on Pytorch models. I'll let you know if I find some more ways to speed things up.

In any case, I'll close this issue for now.
